nltk/translate/chrf_score.py

```diff
@@ -110,28 +132,59 @@ def corpus_chrf(list_of_references, hypotheses, min_len=1, max_len=6, beta=3.0):
     :type max_len: int
     :param beta: the parameter to assign more importance to recall over precision
     :type beta: float
+    :param ignore_whitespace: ignore whitespace characters in scoring
+    :type ignore_whitespace: bool
     :return: the sentence level CHRF score.
     :rtype: float
     """
 
-    assert len(list_of_references) == len(hypotheses), "The number of hypotheses and their references should be the same"
+    assert len(references) == len(hypotheses), (
+        "The number of hypotheses and their references should be the same")
+
+    # Keep f-scores for each n-gram order separate
+    n_gram_fscores = {k: [] for k in range(min_len, max_len + 1)}
+
+    num_sents = 0  # we don't call len() in case these are generators
 
     # Iterate through each hypothesis and their corresponding references.
-    for reference, hypothesis in zip(list_of_references, hypotheses):
-        # Cheating condition to allow users to input strings instead of tokens.
-        if type(reference) and type(hypothesis) != str:
-            reference, hypothesis = ' '.join(reference), ' '.join(hypothesis)
-        # For each order of ngram, calculate the no. of ngram matches and
-        # keep track of no. of ngram in references.
-        ref_ngrams = Counter(everygrams(reference, min_len, max_len))
-        hyp_ngrams = Counter(everygrams(hypothesis, min_len, max_len))
-        overlap_ngrams = ref_ngrams & hyp_ngrams
-        tp = sum(overlap_ngrams.values()) # True positives.
-        tpfp = sum(hyp_ngrams.values()) # True positives + False positives.
-        tffn = sum(ref_ngrams.values()) # True posities + False negatives.
-
-    precision = tp / tpfp
-    recall = tp / tffn
-    factor = beta**2
-    score = (1+ factor ) * (precision * recall) / ( factor * precision + recall)
-    return score
+    for reference, hypothesis in zip(references, hypotheses):
+        num_sents += 1
+
+        # preprocess both reference and hypothesis
+        reference = _preprocess(reference, ignore_whitespace)
+        hypothesis = _preprocess(hypothesis, ignore_whitespace)
+
+        # Calculate f-scores for each sentence and for each n-gram order
+        # separately.
+        for order in range(min_len, max_len + 1):
+            ref_ngrams = Counter(ngrams(reference, order))
+            hyp_ngrams = Counter(ngrams(hypothesis, order))
+
+            # calculate the number of ngram matches
+            overlap_ngrams = ref_ngrams & hyp_ngrams
+            tp = sum(overlap_ngrams.values())  # True positives.
+            tpfp = sum(hyp_ngrams.values())  # True positives + False positives.
+            tffn = sum(ref_ngrams.values())  # True posities + False negatives.
+
+            # Avoid division by zero errors. We also return 0 when both
+            # reference and hypothesis are empty.
+            if 0 in (tp, tpfp, tffn):
+                n_gram_fscores[order].append(0.0)
```

Comment:
Hmmm, I think there's a more elegant way to handle this with try-except to be more explicit about the `ZeroDivisionError`. Also checking for `tp == 0` is unnecessary because it's in the numerator and it won't get into the math domain error.

Reply:
Checking for `tp == 0` was needed because if it's zero, then both precision and recall are zero, which means the denominator of f-score formula is also zero.

But I changed the code to explicitly handle `ZeroDivisionError` so it doesn't matter now :)